File: Striver/Recursion/q23.py
# m - colouring graph problem
# m = number of colours
# n = number of vertices
# graph[][] = adjacency matrix
# all adjacent vertices should have different colours
# return true if graph can be coloured with m colours else false

def isSafe(graph, colour, c, v):
    for i in range(len(graph)):
        if graph[v][i] == 1 and colour[i] == c:
            return False
    return True

def graphColouring(graph, m, colour, v):
    if v == len(graph):
        return True
    for c in range(1, m+1):
        if isSafe(graph, colour, c, v):
            colour[v] = c
            if graphColouring(graph, m, colour, v+1):
                return True
            colour[v] = 0
    return False

graph = [[0, 1, 1, 1],
            [1, 0, 1, 0],
            [1, 1, 0, 1],
            [1, 0, 1, 0]]

m = 3
colour = [0] * len(graph)
print(graphColouring(graph, m, colour, 0))

File: Striver/Recursion/q23.java
package Striver.Recursion;

public class q23 {
    // m - colouring graph problem
    // m = number of colours
    // n = number of vertices
    // graph[][] = adjacency matrix
    // all adjacent vertices should have different colours
    // return true if graph can be coloured with m colours else false

    public static void main(String[] args) {
        int[][] graph = {
                {0, 1, 1, 1},
                {1, 0, 1, 0},
                {1, 1, 0, 1},
                {1, 0, 1, 0}
        };
        int m = 3;
        int n = 4;
        int[] colors = new int[n];
        for (int i = 0; i < n; i++) {
            colors[i] = 0;
        }
        if (solveMColoring(graph, m, n, colors, 0)) {
            System.out.println("Graph can be coloured with " + m + " colours");
        } else {
            System.out.println("Graph cannot be coloured with " + m + " colours");
        }
    }

    public static boolean solveMColoring(int[][] graph, int m, int n, int[] colors, int v) {
        if (v == n) {
            return true;
        }
        for (int i = 1; i <= m; i++) {
            if (isSafe(graph, m, n, colors, v, i)) {
                colors[v] = i;
                if (solveMColoring(graph, m, n, colors, v + 1)) {
                    return true;
                }
                colors[v] = 0;
            }
        }
        return false;
    }

    public static boolean isSafe(int[][] graph, int m, int n, int[] colors, int v, int c) {
        for (int i = 0; i < n; i++) {
            if (graph[v][i] == 1 && colors[i] == c) {
                return false;
            }
        }
        return true;
    }
}
